redactevent.go

```diff
@@ -0,0 +1,131 @@
+	return []byte(r), nil
+}
+
+// UnmarshalJSON impelements the json.Unmarshaller interface using a pointer receiver.
+func (r *rawJSON) UnmarshalJSON(data []byte) error {
+	*r = rawJSON(data)
+	return nil
+}
+
+// RedactEvent strips the user controlled fields from an event, but leaves the
+// fields necessary for authenticating the event.
+func RedactEvent(eventJSON []byte) ([]byte, error) {
+
+	// createContent keeps the fields needed in a m.room.create event.
+	// Create events need to keep the creator.
+	// (In an ideal world they would keep the m.federate flag see matrix-org/synapse#1831)
+	type createContent struct {
+		Creator rawJSON `json:"creator,omitempty"`
+	}
+
+	// joinRulesContent keeps the fields needed in a m.room.join_rules event.
+	// Join rules events need to keep the join_rule key.
+	type joinRulesContent struct {
+		JoinRule rawJSON `json:"join_rule,omitempty"`
+	}
+
+	// powerLevelContent keeps the fields needed in a m.room.power_levels event.
+	// Power level events need to keep all the levels.
+	type powerLevelContent struct {
+		Users         rawJSON `json:"users,omitempty"`
+		UsersDefault  rawJSON `json:"users_default,omitempty"`
+		Events        rawJSON `json:"events,omitempty"`
+		EventsDefault rawJSON `json:"events_default,omitempty"`
+		StateDefault  rawJSON `json:"state_default,omitempty"`
+		Ban           rawJSON `json:"ban,omitempty"`
+		Kick          rawJSON `json:"kick,omitempty"`
+		Redact        rawJSON `json:"redact,omitempty"`
+	}
+
+	// memberContent keeps the fields needed in a m.room.member event.
+	// Member events keep the membership.
+	// (In an ideal world they would keep the third_party_invite see matrix-org/synapse#1831)
+	type memberContent struct {
+		Membership rawJSON `json:"membership,omitempty"`
+	}
+
+	// aliasesContent keeps the fields needed in a m.room.aliases event.
+	// Alias events probably don't need to keep the aliases key, but we need to match synapse here.
+	type aliasesContent struct {
+		Aliases rawJSON `json:"aliases,omitempty"`
+	}
+
+	// historyVisibilityContent keeps the fields needed in a m.room.history_visibility event
+	// History visibility events need to keep the history_visibility key.
+	type historyVisibilityContent struct {
+		HistoryVisibility rawJSON `json:"history_visibility,omitempty"`
+	}
+
+	// allContent keeps the union of all the content fields needed across all the event types.
+	//All the content JSON keys we are keeping are distinct across the different event types.
+	type allContent struct {
+		createContent
+		joinRulesContent
+		powerLevelContent
+		memberContent
+		aliasesContent
+		historyVisibilityContent
+	}
+
+	// eventFields keeps the top level keys needed by all event types.
+	// (In an ideal world they would include the "redacts" key for m.room.redaction events, see matrix-org/synapse#1831)
+	type eventFields struct {
+		EventID        rawJSON    `json:"event_id,omitempty"`
+		Sender         rawJSON    `json:"sender,omitempty"`
```

Comment:
http://matrix.org/docs/spec/client_server/r0.2.0.html#redactions says `user_id` not `sender`. Please cite a Synapse line to prove otherwise.